### ISSUE TITLE
Support Tomcat's secure flag?

### ISSUE THREAD

**Issue by mmoayyed** [spring-projects/spring-boot#5379]
Based on Tomcat's docs:

> Set this attribute to true if you wish to have calls to request.isSecure() to return true for requests received by this Connector. You would want this on an SSL Connector or a non SSL connector that is receiving data from a SSL accelerator, like a crypto card, a SSL appliance or even a webserver. The default value is false.

I couldn't find a property that would allow me to configure this. Is this supported, and if not, could it be added?

**Comment by mmoayyed** [comment]
Happy to issue a PR by the way.

**Comment by philwebb** [comment]
I'm not sure if we need a dedicated property or not. You should be able to do it currently using a `TomcatConnectorCustomizer`.

**Comment by mmoayyed** [comment]
That would require code correct? A prop offered by boot sounds much simpler.

**Comment by snicoll** [comment]
We can't possibly add properties for everything. If the code is simple enough (and I think it is) there's no reason why we should provide a property for that. A general property for all servers we support is probably a different story.

**Comment by mmoayyed** [comment]
I understand. This might be a special case because it's something I imagine most would be interested in, much like the other tomcat-specific settings in proxied deployments of the app. 

You should consider that my target user audience generally has a hard time understanding what an ending XML tag is! So, "you can code it yourself because it's so simple" will never work. I'd basically end up extending boot to provide that sort of property (and everything else that might be needed), so the user could set that property, and it would be best if I didnt have to carry all that change.

**Comment by mmoayyed** [comment]
I'd also emphasize that perhaps not everything, but for most everything, there could be a property. I see no reason why that can't be done. I don't think you should add support for everything all at once and call it a day, but gradually, through requests such as these when there is demand :)

**Comment by snicoll** [comment]
> You should consider that my target user audience generally has a hard time understanding what an ending XML tag is! So, "you can code it yourself because it's so simple" will never work. I

I hope we can agree that the fact your audience is "not able" to write a line of code is not an argument here. If that's required on your side, nothing prevents you to expose a property in your own solution.

> but for most everything, there could be a property

I disagree and the team gave that feedback in plenty of issues. That's programming by properties and we don't want that. We're more than happy to investigate on a case-by-case basis tough and maybe this one will end up in a supported property.

**Comment by mmoayyed** [comment]
> I hope we can agree that the fact your audience is "not able" to write a line of code is not an argument here. If that's required on your side, nothing prevents you to expose a property in your own solution.

Maybe, but more so, I see no reason why they should, or I should, write this on my own. Perhaps this is too presumptuous of me, but I do tend to think this particular case is more common than some obscure tomcat setting. As I said, most proxied or load balanced deployments of a given app will require this property, specially if one is to use the embedded tomcat which we are trying to encourage and recommend. Options between an external and embedded instance of a container need to parallel somewhere, otherwise there is not much use in using the embedded without extra code by every project. 

I would be more than glad to be convinced otherwise, but we have had to set this property on an external tomcat instance, every time. As such, I don't see this as an advanced remote edge case. 

Unless you meant to have the embedded tomcat serve up developers for convenience during dev and testing, in which case, fair point :)  I'll move on to write this on my own.

**Comment by wilkinsona** [comment]
Generally speaking, we prefer to only expose `server.*` configuration properties for settings that are supported across all three containers (Jetty, Tomcat, and Undertow).

Tomcat provides a method on its connector to cause every request to be marked as secure. Boot sets it automatically when Tomcat is configured to use SSL. It also provides `RemoteIpValve` that will mark the request as secure based on the value of the protocol header. Boot configures `RemoteIpValve` when `server.use-forward-headers` is set to `true` and, by default, uses `X-Forwarded-Proto` as the protocol header.

Undertow doesn't have a direct equivalent of `connector.setSecure(true)`. It does, however, provide [`MarkSecureHandler`](https://github.com/undertow-io/undertow/blob/master/servlet/src/main/java/io/undertow/servlet/handlers/MarkSecureHandler.java) which, when configured, has the same effect. Wildfly [uses it](https://github.com/wildfly/wildfly/commit/fe0ed6d6ea63f983b39c11d7e2a99566bdbf60bf). Undertow also has `ProxyPeerAddressHandler` that will (indirectly) mark the request as secure based on the `X-Forwarded-Proto` header. Boot configures `ProxyPeerAddressHandler` when `server.use-forward-headers` is set to `true`.

Jetty doesn't have any equivalent of `connector.setSecure(true)` that I can find. It does have `ForwardedRequestCustomizer` that will mark the request as secure based on the `X-Forwarded-Proto` header. Boot configures `ForwardedRequestCustomizer` when `server.use-forward-headers` is set to `true`.

To summarise, the only mechanism that's supported across all three containers for marking a request as secure is via the use of the `X-Forwarded-Proto` header. Boot already provides support for enabling it via the `server.use-forward-headers` property.

Our recommendation is to use `server.use-forward-headers` and configure whatever is running in from of your Spring Boot instance to include the appropriate headers when forwarding the request. If you want requests to be marked as secure in the absence of those headers then you should use a `TomcatConnectorCustomizer` to do so.

**Comment by mmoayyed** [comment]
Appreciate the feedback. I'll experiment with `server.use-forward-headers`.
